### PR TITLE
Completion support

### PR DESCRIPTION
**Pull request by jchapuis**
This PR contains a `CompletionSupport` trait which can be mixed-in to enable completion for a grammar, i.e. parsers which are 'augmented' with a `completions` method returning possible entry completions for a certain input. This can be used to elaborate as-you-type completions menus or tab-completion experiences, and is e.g. easy to plug with readline to implement a console application. This is ideal to develop e.g. search query grammars, since the flexibility granted by parser-combinators is very valuable in such a context.

All completion-related types, combinator overloads and tests are located in a separate `completions` package and there no impact to existing code. As a result, a small amount of duplication was required for some combinators (but it's very limited).

Note that in addition to this base set of definitions, I could also add a fuzzy completion mix-in to this PR, which allows defining as-you-type experiences supporting fuzzy string matching in an easy way. 

Also, let me know of any required adaptation to the copyright notice for inclusion consideration. Thanks!

### CODE REVIEW

**Comment by gourlaysama** [comment]
Thanks for working on this @jchapuis. That looks like a great addition to parser-combinators! I'll try and review and comment soon :)

**Comment by SethTisue** [comment]
I asked on Gitter (scala/contributors) to see if anyone else would like to help review this.

**Comment by jchapuis** [comment]
Hi all, if there is any way I can help in making this move forward, let me know! This extension has proven very valuable for us, and I find that it is a great application of parser combinators. Indeed, when developing interactive and dynamic grammars, flexibility is much more important than performance - and flexibility is where combinators shine. Completions with dynamic grammars and fuzzy string matching allow for slick natural language search experiences for instance (which is what we use it for).

File: shared/src/test/scala/scala/util/parsing/combinator/completion/CompletionForIntoTest.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  @Test
+  def intoParserWithoutSuccessCompletesToParser(): Unit = {
+    val completions = TestParser.completeString(TestParser.test, "")
+    Assert.assertEquals(Seq(animal, machine), completions)
+  }
+
+  @Test
+  def intoParserWithSuccessCompletesResultingParser(): Unit = {
+    val completions = TestParser.completeString(TestParser.test, animal)
+    Assert.assertEquals(Seq(bear, lion), completions)
```
Comment (olafurpg) [inline shared/src/test/scala/scala/util/parsing/combinator/completion/CompletionForIntoTest.scala]: This is cool! Would it be possible to use completions to generate helpful parse error messages? For example,

```
Parse error!
animal plane
       ^
found: plane
expected: bear | lion
```

Comment (jchapuis) [inline shared/src/test/scala/scala/util/parsing/combinator/completion/CompletionForIntoTest.scala]: @olafurpg thanks! This is a great idea, I will look into this. It should be indeed feasible to "augment" error messages for parsers with CompletionSupport.

**Comment by jchapuis** [comment]
For information, I've made this available in a separate library: https://github.com/jchapuis/scala-parser-combinators-completion. It works in the same way, mixing-in the `CompletionSupport` trait augments parsers with completion support.

So I'm closing this pull request for now (let's see if that library gains some traction and interest).

**Pull request closed without merging**